### PR TITLE
[SPARK-15285][SQL] Generated SpecificSafeProjection.apply method grows beyond 64 KB

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR splits the generated code for `SafeProjection.apply` by using `ctx.splitExpressions()`. This is because the large code body for `NewInstance` may grow beyond 64KB bytecode size for `apply()` method.

Here is [the original PR](https://github.com/apache/spark/pull/13243) for SPARK-15285. However, it breaks a build with Scala 2.10 since Scala 2.10 does not a case class with large number of members. Thus, it was reverted by [this commit](https://github.com/apache/spark/commit/fa244e5a90690d6a31be50f2aa203ae1a2e9a1cf).
## How was this patch tested?

Added new tests by using `DefinedByConstructorParams` instead of case class for scala-2.10

### CODE REVIEW

**Comment by kiszk** [comment]
@cloud-fan @rxin Is this test case what you suggested? I would appreciate your comment.

**Comment by cloud-fan** [comment]
can you add a link to the original PR? thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameComplexTypeSuite.scala [hunk 1]
```diff
@@ -58,4 +59,43 @@ class DataFrameComplexTypeSuite extends QueryTest with SharedSQLContext {
...
+  val s65: String = "65", val s66: String = "66", val s67: String = "67", val s68: String = "68",
+  val s69: String = "69", val s70: String = "70", val s71: String = "71", val s72: String = "72",
+  val s73: String = "73", val s74: String = "74", val s75: String = "75", val s76: String = "76",
+  val s77: String = "77", val s78: String = "78", val s79: String = "79", val s80: String = "80",
+  val s81: String = "81", val s82: String = "82", val s83: String = "83", val s84: String = "84",
+  val s85: String = "85", val s86: String = "86", val s87: String = "87", val s88: String = "88",
+  val s89: String = "89", val s90: String = "90", val s91: String = "91", val s92: String = "92",
+  val s93: String = "93", val s94: String = "94", val s95: String = "95", val s96: String = "96",
+  val s97: String = "97", val s98: String = "98", val s99: String = "99", val s100: String = "100")
+extends DefinedByConstructorParams {}
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameComplexTypeSuite.scala]: remove the useless `{}`

**Comment by cloud-fan** [comment]
thanks, merging to master and 2.0!

**Pull request closed without merging**